packages/edge-preview-authenticated-proxy/src/index.ts

```diff
@@ -0,0 +1,302 @@
+	 * This endpoint takes in the URL parameters:
+	 * 	- `token`   The preview token to authenticate future preview requests
+	 * 						  Crucially, this is _different_ from the session token obtained above
+	 *  - `remote`  Which endpoint to hit with preview requests
+	 *              This should be the workers.dev deployment for the current worker
+	 *  - `prewarm` A fire-and-forget prewarm endpoint to hit to start up the preview
+	 *  - `suffix`  (optional) The pathname + search to hit on the preview worker once redirected
+	 *
+	 * It must be called with a random subdomain (i.e. some-random-data.preview.devprod.cloudflare.dev)
+	 * to provide cookie isolation for the preview.
+	 *
+	 * It will redirect to the suffix provide, setting a cookie with the `token` and `remote`
+	 * for future use.
+	 */
+	if (
+		request.method === "GET" &&
+		url.hostname.endsWith("preview.devprod.cloudflare.dev") &&
+		url.pathname === "/.update-preview-token"
+	) {
+		return updatePreviewToken(url, ctx);
+	}
+
+	/**
+	 * Given a preview token, this endpoint allows for raw http calls to be inspected
+	 * It must be called with a random subdomain (i.e. some-random-data.rawhttp.devprod.cloudflare.dev)
+	 * for consistency with the preview endpoint. This is not currently used, but may be in future
+	 *
+	 * It required two parameters, passed as headers:
+	 *  - `X-CF-Token`  A preview token, as in /.update-preview-token
+	 *  - `X-CF-Remote` Which endpoint to hit with preview requests, as in /.update-preview-token
+	 */
+	if (url.hostname.endsWith("rawhttp.devprod.cloudflare.dev")) {
+		return handleRawHttp(request, url);
+	}
+
+	/**
+	 * Finally, if no other conditions are met, make a preview request to the worker
+	 * This must be called with the same subdomain as used in /.update-preview-token
+	 * so that the cookie will be present. It will swap the host and inject the preview token
+	 * but otherwise will pass the request through unchanged
+	 */
+	const parsedCookies = cookie.parse(request.headers.get("Cookie") ?? "");
+	const { token, remote } = JSON.parse(parsedCookies?.token ?? "{}");
+	if (!token || !remote) {
+		throw new PreviewRequestFailed();
+	}
+
+	const workerUrl = new URL(url);
+	const remoteUrl = new URL(remote);
+	workerUrl.hostname = remoteUrl.hostname;
+	workerUrl.protocol = remoteUrl.protocol;
+
+	return fetch(workerUrl, {
+		...request,
+		headers: {
+			...request.headers,
+			"cf-workers-preview-token": token,
+		},
+	});
+}
+
+// No ecosystem routers support hostname matching 😥
+export default {
+	async fetch(
+		request: Request,
+		env: Env,
+		ctx: ExecutionContext
+	): Promise<Response> {
+		const sentry = new Toucan({
+			dsn: env.SENTRY_DSN,
+			context: ctx,
+			request,
+			requestDataOptions: {
+				allowedHeaders: [
+					"user-agent",
+					"accept-encoding",
+					"accept-language",
+					"cf-ray",
+					"content-length",
+					"content-type",
+					"host",
+				],
+			},
+		});
+		try {
+			return handleRequest(request, ctx);
```

Comment:
OOC I know this makes some subtle difference but I cannot remember what...
Returning a promise to a value directly rather than awaiting the promise and returning the value wrapped in a new promise...

Reply:
Oh wait 🤦 I've added it